buildSrc/src/main/kotlin/com/saveourtool/save/buildutils/JacocoConfiguration.kt

```diff
@@ -49,6 +43,19 @@ fun Project.configureJacoco() {
             html.required.set(true)
         }
     }
+
+    // `application` plugin creates jacocoTestReport task in plugin section (this is definitely incorrect behavior)
+    // AFTER that in "com.saveourtool.save.buildutils.kotlin-library" we try to register this task once again and fail
+    // so the order of plugins in `apply` is critically important
+    val jacocoTestReportTask = if (project.name == "save-cli") {
```

Comment:
@petertrr any ideas why this can happen? I am completely lost :(
```

* What went wrong:
An exception occurred applying plugin request [id: 'com.saveourtool.save.buildutils.kotlin-library']
> Failed to apply plugin 'com.saveourtool.save.buildutils.kotlin-library'.
   > Extension of type 'JacocoReport' does not exist. Currently registered extension types: [ExtraPropertiesExtension]

```

Also funny thing - that plugins order is important in our logic, because we cannot manipulate with the creation of `jacocoTestReport` in `application` plugin

Reply:
> `application` plugin creates jacocoTestReport task in plugin section (this is definitely incorrect behavior)

Why? Gradle calls plugins `apply` method and the plugin performs all its necessary setup, which includes registering of tasks.

> `> Extension of type 'JacocoReport' does not exist. Currently registered extension types: [ExtraPropertiesExtension]`

Looks like jacoco plugin hasn't been applied yet. Do you get this error in branch with `register` for `jacocoTestReport` task? Then may be we pass `configure` incorrectly and it's not lazily evaluated?

Reply:
>>> Do you get this error in branch with register for jacocoTestReport task?
Nope, I get it on `save-cli` (on the branch where we use jacocoTestReport task created in `application` plugin)

Reply:
`jacoco` plugin reacts to the presence of `java` plugin (which is applied by `application` plugin) as described [here](https://docs.gradle.org/current/userguide/jacoco_plugin.html#sec:jacoco_tasks). I think `getByName` is an incorrect choice [after all](https://docs.gradle.org/current/userguide/task_configuration_avoidance.html#sec:how_do_i_reference_a_task), because it'll get task immediately and call `configure` on the task instead of a `TaskProvider`. I think more appropriate would be to use `tasks.named` and then call `configure` on the returned provider.

Reply:
[How do I reference a task without creating/configuring it?](https://docs.gradle.org/current/userguide/task_configuration_avoidance.html#sec:how_do_i_reference_a_task)

Instead of referencing a task object, you can work with a registered task via a [TaskProvider](https://docs.gradle.org/current/javadoc/org/gradle/api/tasks/TaskProvider.html) object. A [TaskProvider](https://docs.gradle.org/current/javadoc/org/gradle/api/tasks/TaskProvider.html) can be obtained in several ways including when calling [TaskContainer.register(java.lang.String)](https://docs.gradle.org/current/javadoc/org/gradle/api/tasks/TaskContainer.html#register-java.lang.String-) or using the [TaskCollection.named(java.lang.String)](https://docs.gradle.org/current/javadoc/org/gradle/api/tasks/TaskCollection.html#named-java.lang.String-) method.

Yes, this is what they suggest

Reply:
https://docs.gradle.org/current/userguide/task_configuration_avoidance.html#sec:old_vs_new_configuration_api_overview

Reply:
If you are configuring a task by name, you will need to use the configuration avoidance equivalent. See the [table below](https://docs.gradle.org/current/userguide/task_configuration_avoidance.html#sec:old_vs_new_configuration_api_overview) to identify the alternative.

Reply:
In TaskProvider they have [configure](https://docs.gradle.org/current/javadoc/org/gradle/api/tasks/TaskProvider.html#configure-org.gradle.api.Action-)​([Action](https://docs.gradle.org/current/javadoc/org/gradle/api/Action.html)<? super [T](https://docs.gradle.org/current/javadoc/org/gradle/api/tasks/TaskProvider.html)> action)

Reply:
But `register` returns `TaskProvider` too, so using `named` should only make two branches more similar

Reply:
yeah, but that API without examples was very confusing (or may be I am too stupid for gradle)